### DISCUSSION TITLE
Import a local nuget build (ARM UWP)

### DISCUSSION THREAD

**Discussion by austinbhale** [StereoKit/StereoKit#499] [Q&A] [answered]
Is there a way to build the NuGet package and reference it locally?

I want to make this line public since I'm returning a mesh from my C++ DLL:
https://github.com/StereoKit/StereoKit/blob/f4922af7a33f8314f742c1cc3331dcd66e062f71/StereoKit/Assets/Mesh.cs#L72 

#### What I've tried
Successfully built the `Build-Nuget.ps1` shell script with Android and Linux disabled. Whenever I try to import in the NuGet package manager, I receive the following message:
`NU1101: Unable to find package StereoKitC. No packages exist with this id in source(s): External, nuget.org`

Is there a way to build the local NuGet package without the StereoKitC dependency? I noticed the preview build release page (https://www.nuget.org/packages/StereoKit/0.3.7-preview.5#dependencies-body-tab) doesn't have StereoKitC listed as a dependency.

**Comment by maluoi** [comment, marked as answer]
There is no StereoKitC package, StereoKitC is just the native DLL that C#'s StereoKit wraps! The local StereoKit NuGet package _must_ have the StereoKitC.dll with it, or it won't be able to PInvoke into it.

The app itself doesn't need to know anything about StereoKitC itself, so you don't need to directly reference that. This is the purpose of the NuGet package's included build scripts (.target files), which are in charge of copying around the correct DLLs to the right folders.

With respect to `internal Mesh(IntPtr mesh)`, I'd be happy to have something like this in the official API, but I'd probably prefer it as a static method rather than as a constructor. Like a `Mesh Mesh.FromNativeSKMesh(IntPtr meshAsset)` or something.

**Reply by austinbhale** [reply]
Exactly! It's strange why the current project configuration lists it as a dependency and even searches online for it.

Here are some steps to replicate this error with the current dev branch:
1. Change `$buildLinux` & `$buildAndroid` to false in .\Build-Nuget.ps1
2. With these changed to false, the script early exits unless I comment out the following Android & Linux libs in `StereoKit.csproj`:
```
<!-- Android Libraries -->
<Content Pack="true" Include="..\bin\distribute\bin\android\arm64-v8a\release\libStereoKitC.so"    PackagePath="runtimes\android-arm64\native" />
<Content Pack="true" Include="..\bin\distribute\bin\android\arm64-v8a\release\StereoKitC.sym"      PackagePath="runtimes\android-arm64\native" />
<Content Pack="true" Include="..\bin\distribute\bin\android\arm64-v8a\release\libopenxr_loader.so" PackagePath="runtimes\android-arm64\native" />

<!-- Linux Libraries -->
<Content Pack="true" Include="..\bin\distribute\bin\linux\x64\release\libStereoKitC.so"   PackagePath="runtimes\linux-x64\native\" />
<Content Pack="true" Include="..\bin\distribute\bin\linux\x64\release\StereoKitC.sym"     PackagePath="runtimes\linux-x64\native\" />
<Content Pack="true" Include="..\bin\distribute\bin\linux\arm64\release\libStereoKitC.so" PackagePath="runtimes\linux-arm64\native\" />
<Content Pack="true" Include="..\bin\distribute\bin\linux\arm64\release\StereoKitC.sym"   PackagePath="runtimes\linux-arm64\native\" />
```
Now, the NuGet has successfully been built. 
3. Create a new project, then try to import the package via package source in the local machine. It receives the following error:
```
  GET https://api.nuget.org/v3-flatcontainer/stereokitc/index.json
  NotFound https://api.nuget.org/v3-flatcontainer/stereokitc/index.json 764ms
NU1101: Unable to find package StereoKitC. No packages exist with this id in source(s): External, nuget.org
Package restore failed. Rolling back package changes for 'StereoKitProject1'.
```

If it helps, I've attached the packaged file here. Curious to see if I'm missing a step here. Thanks so much!

Also, `FromNativeSKMesh` would be wonderful! I could envision something like this for most, if not all, of the private asset types in StereoKitC.
[StereoKit.0.3.7-preview.5.zip](https://github.com/StereoKit/StereoKit/files/9856932/StereoKit.0.3.7-preview.5.zip)

**Reply by maluoi** [reply]
This is basically how I do fast builds myself, and your build does actually work just fine in my setup! It may perhaps be more about how you're referencing the .nuget package? It should be near identical to the way you reference official builds from the NuGet server.

Iteration on a fully built nuget package is generally not a lot of fun, but here's how I do it:
- Set a [NuGet package source](https://learn.microsoft.com/en-us/nuget/consume-packages/install-use-packages-visual-studio#package-sources), I have one set up for StereoKit's bin folder.
- Delete any old version installed in `C:\Users\[user]\.nuget\packages\stereokit`.
- Build the .nuget package.
- Set your VS project's NuGet package source to your custom one.
- Clean rebuild the test project.

**Reply by austinbhale** [reply]
> Delete any old version installed in `C:\Users\[user]\.nuget\packages\stereokit`.

This step was key. Sweet, it works! Thanks for sharing your deployment pipeline - you've got a nice setup here.
